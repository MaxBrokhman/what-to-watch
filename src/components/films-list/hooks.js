import {useState, useCallback, useEffect} from 'react';

const TIMER = 1000;
const START_LIMIT = 8;
const SHOW_MORE_STEP = 20;

export const useVideoTimer = () => {
  const [activeCard, setActiveCard] = useState(null);
  const [timer, setTimer] = useState(null);
  const hoverHandler = (movie) => () => {
    setTimer(setTimeout(() => {
      setActiveCard(movie);
    }, TIMER));
  };
  const leaveHandler = () => {
    clearTimeout(timer);
    setActiveCard(null);
  };

  return {
    activeCard,
    hoverHandler,
    leaveHandler,
  };
};

export const useFilteredFilms = (genre, movies) => {
  const [limit, setLimit] = useState(START_LIMIT);

  useEffect(() => {
    setLimit(START_LIMIT);
  }, [genre]);

  const filteredMovies = genre === `all genres`
    ? movies
    : movies.filter((film) => film.genre === genre);

  const areAllMoviesLoaded = limit >= filteredMovies.length;

  const moviesLimit = areAllMoviesLoaded
    ? filteredMovies.length
    : limit;

  const showMoreClickHandler = useCallback(() => {
    setLimit(limit + SHOW_MORE_STEP);
  }, [limit]);
  return {
    movies: filteredMovies.slice(0, moviesLimit),
    showMoreClickHandler,
    areAllMoviesLoaded,
  };
};
