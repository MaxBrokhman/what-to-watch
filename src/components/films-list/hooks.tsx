import {
  useState,
  useCallback,
  useEffect,
} from 'react';

import {TFilm} from '../../reducer/types';

const TIMER = 1000;
const START_LIMIT = 8;
const SHOW_MORE_STEP = 20;

type TUseVideoTimer = {
  activeCard: TFilm;
  hoverHandler: (movie: TFilm) => () => void;
  leaveHandler: () => void;
}

export const useVideoTimer = (): TUseVideoTimer => {
  const [activeCard, setActiveCard] = useState(null);
  const [timer, setTimer] = useState(null);
  const hoverHandler = (movie) => (): void => {
    setTimer(setTimeout(() => {
      setActiveCard(movie);
    }, TIMER));
  };
  const leaveHandler = (): void => {
    clearTimeout(timer);
    setActiveCard(null);
  };

  useEffect(() => {
    return (): void => {
      clearTimeout(timer);
      setActiveCard(null);
    };
  }, [timer]);

  return {
    activeCard,
    hoverHandler,
    leaveHandler,
  };
};

type TUseFilteredFilms = {
  movies: Array<TFilm>;
  areAllMoviesLoaded: boolean;
  showMoreClickHandler: () => void;
}

export const useFilteredFilms = (genre: string, movies: Array<TFilm>): TUseFilteredFilms => {
  const [limit, setLimit] = useState(START_LIMIT);

  useEffect(() => {
    setLimit(START_LIMIT);
  }, [genre]);

  const filteredMovies = genre === `all genres`
    ? movies
    : movies.filter((film) => film.genre === genre);

  const areAllMoviesLoaded = limit >= filteredMovies.length;

  const moviesLimit = areAllMoviesLoaded
    ? filteredMovies.length
    : limit;

  const showMoreClickHandler = useCallback(() => {
    setLimit(limit + SHOW_MORE_STEP);
  }, [limit]);
  return {
    movies: filteredMovies.slice(0, moviesLimit),
    showMoreClickHandler,
    areAllMoviesLoaded,
  };
};
