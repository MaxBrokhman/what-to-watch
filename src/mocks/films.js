/* eslint-disable */
export const films = [
  {
    name: `Fantastic Beasts: The Crimes of Grindelwald`,
    preview_image: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    genre: `fantasy`,
    background_image: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    poster_image: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    run_time: 88,
    released: 2003,
    id: 1,
  },
  {
    name: `Bohemian Rhapsody`,
    preview_image: `img/bohemian-rhapsody.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    genre: `drama`,
    background_image: `img/bohemian-rhapsody.jpg`,
    poster_image: `img/bohemian-rhapsody.jpg`,
    run_time: 88,
    released: 2003,
    id: 2,
  },
  {
    name: `Macbeth`,
    preview_image: `img/macbeth.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    genre: `drama`,
    background_image: `img/macbeth.jpg`,
    poster_image: `img/macbeth.jpg`,
    run_time: 88,
    released: 2003,
    id: 3,
  },
  {
    name: `Aviator`,
    preview_image: `img/aviator.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    genre: `thriller`,
    background_image: `img/aviator.jpg`,
    poster_image: `img/aviator.jpg`,
    run_time: 88,
    released: 2003,
    id: 4,
  },
  {
    name: `We need to talk about Kevin`,
    preview_image: `img/we-need-to-talk-about-kevin.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    genre: `comedy`,
    background_image: `img/we-need-to-talk-about-kevin.jpg`,
    poster_image: `img/we-need-to-talk-about-kevin.jpg`,
    run_time: 88,
    released: 2003,
    id: 5
  }
];
