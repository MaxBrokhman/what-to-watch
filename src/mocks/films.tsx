/* eslint-disable */
import {TFilm} from "../reducer/types";

export const films: Array<TFilm> = [
  {
    name: `Fantastic Beasts: The Crimes of Grindelwald`,
    preview_image: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    genre: `fantasy`,
    background_image: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    poster_image: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    run_time: 88,
    released: 2003,
    id: 1,
    director: ``,
    starring: [],
    rating: 9,
    description: ``,
  },
  {
    name: `Bohemian Rhapsody`,
    preview_image: `img/bohemian-rhapsody.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    genre: `drama`,
    background_image: `img/bohemian-rhapsody.jpg`,
    poster_image: `img/bohemian-rhapsody.jpg`,
    run_time: 88,
    released: 2003,
    id: 2,
    director: ``,
    starring: [],
    rating: 9,
    description: ``,
  },
  {
    name: `Macbeth`,
    preview_image: `img/macbeth.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    genre: `drama`,
    background_image: `img/macbeth.jpg`,
    poster_image: `img/macbeth.jpg`,
    run_time: 88,
    released: 2003,
    id: 3,
    director: ``,
    starring: [],
    rating: 9,
    description: ``,
  },
  {
    name: `Aviator`,
    preview_image: `img/aviator.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    genre: `thriller`,
    background_image: `img/aviator.jpg`,
    poster_image: `img/aviator.jpg`,
    run_time: 88,
    released: 2003,
    id: 4,
    director: ``,
    starring: [],
    rating: 9,
    description: ``,
  },
  {
    name: `We need to talk about Kevin`,
    preview_image: `img/we-need-to-talk-about-kevin.jpg`,
    preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    genre: `comedy`,
    background_image: `img/we-need-to-talk-about-kevin.jpg`,
    poster_image: `img/we-need-to-talk-about-kevin.jpg`,
    run_time: 88,
    released: 2003,
    id: 5,
    director: ``,
    starring: [],
    rating: 9,
    description: ``,
  }
];
