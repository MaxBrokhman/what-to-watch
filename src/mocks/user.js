export const user = {
  id: 1,
  email: `Oliver.conner@gmail.com`,
  name: `Oliver`,
  // eslint-disable-next-line
  avatar_url: `#`,
};
