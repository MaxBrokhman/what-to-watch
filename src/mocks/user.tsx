import {TUser} from "../reducer/types";

export const user: TUser = {
  email: `Oliver.conner@gmail.com`,
  name: `Oliver`,
  // eslint-disable-next-line
  avatar_url: `#`,
};
