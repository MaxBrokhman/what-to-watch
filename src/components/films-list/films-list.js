import React, {Fragment} from 'react';

import {MovieCard} from '../movie-card/movie-card';
import {useVideoTimer, useFilteredFilms} from './hooks';
import {useAppContext} from '../../reducer/reducer';

export const FilmsList = () => {
  const {
    activeCard,
    hoverHandler,
    leaveHandler,
  } = useVideoTimer();
  const {state} = useAppContext();
  const {
    movies,
    showMoreClickHandler,
    areAllMoviesLoaded,
  } = useFilteredFilms(state.genre, state.filmsList);
  return (
    <Fragment>
      <div className="catalog__movies-list">
        {
          movies.map((movie) =>
            <MovieCard
              movie={movie}
              hoverHandler={hoverHandler}
              leaveHandler={leaveHandler}
              activeCard={activeCard}
              key={movie.id}
            />)
        }
      </div>
      {
        !areAllMoviesLoaded && (
          <div className="catalog__more">
            <button
              className="catalog__button"
              type="button"
              onClick={showMoreClickHandler}
            >
            Show more
            </button>
          </div>
        )
      }
    </Fragment>
  );
};
