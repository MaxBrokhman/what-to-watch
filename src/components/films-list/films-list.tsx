import React, {Fragment} from 'react';

import {MovieCard} from '../movie-card/movie-card';
import {useVideoTimer, useLimitedFilms} from './hooks';
import {TFilm} from '../../reducer/types';

type TFilmsListProps = {
  films: Array<TFilm>;
}

export const FilmsList = ({films}: TFilmsListProps): JSX.Element => {
  const {
    activeCard,
    hoverHandler,
    leaveHandler,
  } = useVideoTimer();
  const {
    limitedFilms,
    showMoreClickHandler,
    areAllMoviesLoaded,
  } = useLimitedFilms(films);
  return (
    <Fragment>
      <div className="catalog__movies-list">
        {
          limitedFilms.map((movie) =>
            <MovieCard
              movie={movie}
              hoverHandler={hoverHandler}
              leaveHandler={leaveHandler}
              activeCard={activeCard}
              key={movie.id}
            />)
        }
      </div>
      {
        !areAllMoviesLoaded && (
          <div className="catalog__more">
            <button
              className="catalog__button"
              type="button"
              onClick={showMoreClickHandler}
            >
            Show more
            </button>
          </div>
        )
      }
    </Fragment>
  );
};
