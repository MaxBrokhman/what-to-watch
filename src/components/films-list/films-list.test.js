import React from 'react';
import renderer from 'react-test-renderer';
import {BrowserRouter} from 'react-router-dom';

import {FilmsList} from './films-list';
import {films} from '../../mocks/films';

it(`Renders FilmsList component correctly`, () => {
  const tree = renderer
                .create(
                    <BrowserRouter>
                      <FilmsList genre={`drama`} films={films} />
                    </BrowserRouter>
                )
                .toJSON();
  expect(tree).toMatchSnapshot();
});
