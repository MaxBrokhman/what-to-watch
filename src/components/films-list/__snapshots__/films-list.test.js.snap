// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders FilmsList component correctly 1`] = `
<div
  className="catalog__movies-list"
/>
`;
